File: single_file_demo/jcs.js
/**
 * RFC 8785 JSON Canonicalization Scheme (JCS) — JavaScript port.
 *
 * Byte-identical to ``internal/infrastructure/jcs.py``. The two
 * implementations are cross-verified in ``scripts/verify_cross_runtime.py``
 * against a fixture corpus: every input must produce the same UTF-8 bytes
 * from both runtimes. Any divergence is a correctness regression and is
 * gated in CI before merge.
 *
 * Scope matches the Python side exactly:
 *   object (plain object, string keys)
 *   array
 *   string
 *   integer (finite, !isFloat)
 *   boolean
 *   null
 *
 * Floats are intentionally UNSUPPORTED — the Python impl throws ValueError
 * on them, this impl throws TypeError with the same reasoning: the VC 2.0 +
 * eddsa-jcs-2022 path inside SUM never emits floats (integers and
 * ISO-8601 strings only), so shipping an ES6-ToString serializer for a
 * type we do not use would be untested weight.
 *
 * Key-sort rule (RFC 8785 §3.2.3): UTF-16 code-unit sequence.
 * JavaScript strings ARE UTF-16 natively, so the default lexicographic
 * comparison (Array.prototype.sort with no compareFn) produces exactly
 * the UTF-16-code-unit order. The Python side uses
 * ``s.encode("utf-16-be")`` as its sort key to achieve the same thing on
 * its UTF-8-backed strings. The two approaches yield identical
 * ordering for every Unicode input, verified by the supplementary-
 * character test case (U+1F600 sorts before U+E000).
 *
 * String escaping (RFC 8785 §3.2.2):
 *   "    → \"
 *   \    → \\
 *   \b \t \n \f \r   short-escape forms
 *   other <0x20        \u00XX (LOWERCASE hex)
 *   /                  NOT escaped (RFC 8785 diverges from 8259 here)
 *   everything else    passed through as UTF-8
 *
 * No dependencies. Pure ES2015+. Works in Node ≥ 14 and every modern
 * browser. Intended as the canonicalization substrate for the
 * single-file SUM demo artifact (paste-into-Claude) — provenance IDs
 * and any future signed bundle use it.
 */
'use strict';

/**
 * Canonicalize a JSON-serializable value per RFC 8785.
 *
 * @param {*} obj  supported: object, array, string, integer, boolean, null
 * @returns {Uint8Array}  UTF-8 byte sequence, byte-identical to the
 *                        Python ``canonicalize(obj)`` return.
 */
function canonicalize(obj) {
  const encoder = new TextEncoder();
  return encoder.encode(canonicalizeToStr(obj));
}

/**
 * Canonicalize to a JS string (UTF-16 by default in the engine, but the
 * sequence of code units matches the target canonical form byte-for-byte
 * after UTF-8 re-encoding via canonicalize()).
 *
 * @param {*} obj
 * @returns {string}
 */
function canonicalizeToStr(obj) {
  return _encode(obj);
}

function _encode(obj) {
  if (obj === true) return 'true';
  if (obj === false) return 'false';
  if (obj === null || obj === undefined) return 'null';
  const t = typeof obj;
  if (t === 'string') return _encodeString(obj);
  if (t === 'number') {
    if (!Number.isFinite(obj)) {
      throw new TypeError(
        'JCS: non-finite numbers (NaN, Infinity) are not valid JSON'
      );
    }
    if (!Number.isInteger(obj)) {
      throw new TypeError(
        'JCS: floating-point values are not supported by this implementation ' +
        '(use integer or ISO-8601 string representations instead)'
      );
    }
    return String(obj);
  }
  if (t === 'bigint') {
    return obj.toString();
  }
  if (Array.isArray(obj)) {
    const parts = [];
    for (const item of obj) parts.push(_encode(item));
    return '[' + parts.join(',') + ']';
  }
  if (t === 'object') {
    return _encodeObject(obj);
  }
  throw new TypeError(`JCS: unsupported type ${t}`);
}

function _encodeObject(obj) {
  const keys = Object.keys(obj);
  for (const k of keys) {
    if (typeof k !== 'string') {
      throw new TypeError(`JCS: object keys must be str, got ${typeof k}`);
    }
  }
  // Default string sort in JS is UTF-16 code-unit ordering, which is
  // exactly what RFC 8785 §3.2.3 specifies. No custom compareFn needed.
  keys.sort();
  const parts = [];
  for (const k of keys) {
    parts.push(_encodeString(k) + ':' + _encode(obj[k]));
  }
  return '{' + parts.join(',') + '}';
}

function _encodeString(s) {
  let out = '"';
  for (let i = 0; i < s.length; i++) {
    const cp = s.charCodeAt(i);
    if (cp === 0x22) out += '\\"';
    else if (cp === 0x5C) out += '\\\\';
    else if (cp === 0x08) out += '\\b';
    else if (cp === 0x09) out += '\\t';
    else if (cp === 0x0A) out += '\\n';
    else if (cp === 0x0C) out += '\\f';
    else if (cp === 0x0D) out += '\\r';
    else if (cp < 0x20) {
      out += '\\u' + cp.toString(16).padStart(4, '0');
    } else {
      out += s[i];
    }
  }
  out += '"';
  return out;
}

module.exports = {
  canonicalize,
  canonicalizeToStr,
};

File: internal/infrastructure/jcs.py
"""RFC 8785 JSON Canonicalization Scheme (JCS) — pure-Python implementation.

JCS produces a deterministic UTF-8 byte sequence for a JSON value so that
cryptographic signatures remain stable across implementations. The output is
the byte string that downstream code hashes and signs — no framework, no
external canonicalizer, no C extension.

Scope. The Verifiable Credentials 2.0 + eddsa-jcs-2022 path used by SUM only
emits JSON values of these types:

    dict[str, ...]   object
    list             array
    tuple            array (JSON has no tuple; we coerce)
    str              string
    int / bool       number / boolean
    None             null

RFC 8785 additionally specifies rules for floats (IEEE 754 → ECMAScript
Number.prototype.toString). SUM never emits floats inside a VC credentialSubject
(we use integers and ISO-8601 datetime strings), so this implementation raises
``ValueError`` if a float is encountered rather than ship an untested
serialization path for a type we do not use.

The key-sort rule in RFC 8785 §3.2.3 is "code unit sequence of the UTF-16
representation". ``str.encode("utf-16-be")`` yields exactly that byte order
for sorting, which gives the correct answer for both BMP and supplementary
characters (Python's default string sort differs on supplementary code points).
"""
from __future__ import annotations

from typing import Any, Mapping

__all__ = ["canonicalize", "canonicalize_to_str"]


def canonicalize(obj: Any) -> bytes:
    """Return the RFC 8785 canonical UTF-8 bytes for ``obj``."""
    return _encode(obj).encode("utf-8")


def canonicalize_to_str(obj: Any) -> str:
    """Return the RFC 8785 canonical form as a ``str`` (not yet UTF-8 encoded).

    The JCS byte output is always UTF-8; the str form is equal to
    ``canonicalize(obj).decode('utf-8')``.
    """
    return _encode(obj)


def _encode(obj: Any) -> str:
    if obj is True:
        return "true"
    if obj is False:
        return "false"
    if obj is None:
        return "null"
    if isinstance(obj, str):
        return _encode_string(obj)
    if isinstance(obj, bool):  # pragma: no cover — unreachable; handled above
        return "true" if obj else "false"
    if isinstance(obj, int):
        return str(obj)
    if isinstance(obj, float):
        raise ValueError(
            "JCS: floating-point values are not supported by this implementation "
            "(use integer or ISO-8601 string representations instead)"
        )
    if isinstance(obj, Mapping):
        return _encode_object(obj)
    if isinstance(obj, (list, tuple)):
        return "[" + ",".join(_encode(x) for x in obj) + "]"
    raise TypeError(f"JCS: unsupported type {type(obj).__name__}")


def _encode_object(obj: Mapping[Any, Any]) -> str:
    items = []
    for k in obj.keys():
        if not isinstance(k, str):
            raise TypeError(
                f"JCS: object keys must be str, got {type(k).__name__}"
            )
        items.append(k)
    # RFC 8785 §3.2.3: sort by UTF-16 code unit sequence.
    items.sort(key=lambda s: s.encode("utf-16-be"))
    encoded_pairs = [f"{_encode_string(k)}:{_encode(obj[k])}" for k in items]
    return "{" + ",".join(encoded_pairs) + "}"


def _encode_string(s: str) -> str:
    # RFC 8785 §3.2.2 / RFC 8259 §7. Short escapes for B/T/N/F/R; \uXXXX for
    # every other control character; only " and \ are backslash-escaped beyond
    # that. The solidus ('/') is NOT escaped. Hex digits are lowercase.
    out = ['"']
    for ch in s:
        cp = ord(ch)
        if cp == 0x22:
            out.append("\\\"")
        elif cp == 0x5C:
            out.append("\\\\")
        elif cp == 0x08:
            out.append("\\b")
        elif cp == 0x09:
            out.append("\\t")
        elif cp == 0x0A:
            out.append("\\n")
        elif cp == 0x0C:
            out.append("\\f")
        elif cp == 0x0D:
            out.append("\\r")
        elif cp < 0x20:
            out.append(f"\\u{cp:04x}")
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)
